File: javascript/bob/bob.js
module.exports = function Bob() {
    return {
        hey: (input) => {
            if (input.trim() === '') {
                return 'Fine. Be that way!';
            } else if (input.match(/\b[A-Z0-9]{2,}\b/) && (input === input.toUpperCase() || input.endsWith('!'))) {
                return 'Whoa, chill out!';
            }
            return input.endsWith('?') ? 'Sure.' : 'Whatever.';
        }
    };
}

File: scala/bob/src/main/scala/Bob.scala
class Bob {
  case class Element(str: String, shout: Boolean, end: String)

  def hey(str: String) : String = {
    val pattern = "[A-Za-z]".r;
    Element(str.trim(), str.toUpperCase() == str && pattern.findFirstIn(str) != None, str.takeRight(1)) match {
      case Element("", _, _) =>  "Fine. Be that way!"
      case Element(_, true, _) =>  "Whoa, chill out!"
      case Element(_, _, "?") => "Sure."
      case _ => "Whatever."
    }
  }
}
